File: src/harness.js
import { mkdir, writeFile } from "node:fs/promises";
import { dirname } from "node:path";
import { cloneJson, createTraceId, sha256, stableStringify } from "./hash.js";
import { defaultRedactor } from "./redaction.js";

export class AgentReplayHarness {
  constructor(options = {}) {
    this.mode = options.mode ?? "record";
    this.project = options.project ?? "default";
    this.agent = options.agent ?? {};
    this.trace = options.trace ?? null;
    this.traceId = this.trace?.traceId ?? options.traceId ?? createTraceId();
    this.startedAt = options.startedAt ?? new Date().toISOString();
    this.events = [];
    this.toolManifest = cloneJson(options.toolManifest ?? []);
    this.replayCursor = 0;
    this.replayMismatches = [];
    this.redact = options.redact ?? defaultRedactor;
  }

  recordInput(input) {
    this.events.push({
      id: eventId(this.events.length),
      type: "user_input",
      ts: new Date().toISOString(),
      input: this.redact(cloneJson(input))
    });
  }

  recordFinalOutput(output) {
    this.events.push({
      id: eventId(this.events.length),
      type: "final_output",
      ts: new Date().toISOString(),
      output: this.redact(cloneJson(output))
    });
  }

  wrapTool(name, handler, options = {}) {
    this.#registerTool(name, options);

    return async (args = {}, context = {}) => {
      const started = Date.now();
      const safeArgs = this.redact(cloneJson(args));

      if (this.mode === "replay") {
        return this.#replayToolCall(name, safeArgs, context, started);
      }

      const preState = options.snapshot
        ? this.redact(cloneJson(await options.snapshot(args, context)))
        : undefined;

      let response;
      let error;
      try {
        response = await handler(args, context);
      } catch (caught) {
        error = {
          name: caught.name,
          message: caught.message
        };
      }

      const postState = options.snapshot
        ? this.redact(cloneJson(await options.snapshot(args, context)))
        : undefined;
      const sideEffects = options.diff
        ? this.redact(cloneJson(await options.diff(preState, postState, args, response, context)))
        : diffState(preState, postState);
      const safeResponse = this.redact(cloneJson(response));

      const event = {
        id: eventId(this.events.length),
        type: "tool_call",
        ts: new Date().toISOString(),
        tool: name,
        args: safeArgs,
        argHash: sha256(safeArgs),
        response: safeResponse,
        responseHash: sha256(safeResponse ?? null),
        error,
        approval: context.approval ?? null,
        preState,
        postState,
        sideEffects,
        durationMs: Date.now() - started
      };

      this.events.push(stripUndefined(event));

      if (error) {
        const thrown = new Error(error.message);
        thrown.name = error.name;
        throw thrown;
      }

      return response;
    };
  }

  finalize(options = {}) {
    if (this.mode === "replay") {
      this.#recordMissingReplayToolCalls();
    }

    return stripUndefined({
      schemaVersion: "agentreplay.trace.v1",
      traceId: this.traceId,
      project: this.project,
      agent: this.agent,
      startedAt: this.startedAt,
      endedAt: new Date().toISOString(),
      sourceTraceId: this.trace?.traceId,
      replayMismatches: this.replayMismatches,
      toolManifest: this.toolManifest,
      events: this.events,
      expectedOutcome: options.expectedOutcome,
      metadata: options.metadata
    });
  }

  async save(filePath, options = {}) {
    const trace = this.finalize(options);
    await mkdir(dirname(filePath), { recursive: true });
    await writeFile(filePath, `${JSON.stringify(trace, null, 2)}\n`);
    return trace;
  }

  #replayToolCall(name, safeArgs, context, started) {
    const expectedTools = this.trace.events.filter((event) => event.type === "tool_call");
    const expected = expectedTools[this.replayCursor];
    this.replayCursor += 1;

    if (!expected) {
      const mismatch = {
        kind: "unexpected_tool_call",
        tool: name,
        args: safeArgs
      };
      this.replayMismatches.push(mismatch);
      throw new ReplayMismatchError(mismatch);
    }

    const mismatch = compareExpectedToolCall(expected, name, safeArgs);
    if (mismatch) {
      this.replayMismatches.push(mismatch);
    }

    const event = stripUndefined({
      id: eventId(this.events.length),
      type: "tool_call",
      ts: new Date().toISOString(),
      tool: name,
      args: safeArgs,
      argHash: sha256(safeArgs),
      response: cloneJson(expected.response),
      responseHash: sha256(expected.response ?? null),
      approval: context.approval ?? null,
      preState: cloneJson(expected.preState),
      postState: cloneJson(expected.postState),
      sideEffects: cloneJson(expected.sideEffects),
      replay: {
        expectedTool: expected.tool,
        expectedArgHash: expected.argHash,
        matched: !mismatch
      },
      durationMs: Date.now() - started
    });

    this.events.push(event);

    if (expected.error) {
      const thrown = new Error(expected.error.message);
      thrown.name = expected.error.name;
      throw thrown;
    }

    return cloneJson(expected.response);
  }

  #recordMissingReplayToolCalls() {
    const expectedTools = this.trace.events.filter((event) => event.type === "tool_call");

    for (let index = this.replayCursor; index < expectedTools.length; index += 1) {
      const expected = expectedTools[index];
      const alreadyRecorded = this.replayMismatches.some(
        (mismatch) =>
          mismatch.kind === "missing_expected_tool_call" &&
          mismatch.expectedEventId === expected.id
      );

      if (!alreadyRecorded) {
        this.replayMismatches.push({
          kind: "missing_expected_tool_call",
          expectedEventId: expected.id,
          expectedTool: expected.tool,
          expectedIndex: index
        });
      }
    }
  }

  #registerTool(name, options) {
    if (this.toolManifest.some((tool) => tool.name === name)) {
      return;
    }

    this.toolManifest.push(
      stripUndefined({
        name,
        description: options.description,
        inputSchema: options.inputSchema,
        outputSchema: options.outputSchema,
        hasSnapshot: Boolean(options.snapshot),
        hasDiff: Boolean(options.diff)
      })
    );
  }
}

export class ReplayMismatchError extends Error {
  constructor(mismatch) {
    super(`Replay mismatch: ${mismatch.kind}`);
    this.name = "ReplayMismatchError";
    this.mismatch = mismatch;
  }
}

function compareExpectedToolCall(expected, tool, args) {
  if (expected.tool !== tool) {
    return {
      kind: "tool_name_mismatch",
      expected: expected.tool,
      actual: tool
    };
  }

  if (stableStringify(expected.args) !== stableStringify(args)) {
    return {
      kind: "tool_args_mismatch",
      tool,
      expected: expected.args,
      actual: args
    };
  }

  return null;
}

function diffState(before, after) {
  if (before === undefined || after === undefined) {
    return [];
  }

  if (stableStringify(before) === stableStringify(after)) {
    return [];
  }

  return [
    {
      type: "state_changed",
      beforeHash: sha256(before),
      afterHash: sha256(after)
    }
  ];
}

function eventId(index) {
  return `evt_${String(index + 1).padStart(4, "0")}`;
}

function stripUndefined(value) {
  if (Array.isArray(value)) {
    return value.map(stripUndefined);
  }

  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value)
        .filter(([, nested]) => nested !== undefined)
        .map(([key, nested]) => [key, stripUndefined(nested)])
    );
  }

  return value;
}

File: sdk/go/agentreplay/harness.go
package agentreplay

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"time"
)

var defaultKeyPattern = regexp.MustCompile(`(?i)^(token|secret|password|authorization|apikey|api_key|email|to|from|cc|bcc)$`)
var defaultEmailPattern = regexp.MustCompile(`^[^\s@]+@[^\s@]+\.[^\s@]+$`)

type Harness struct {
	Project      string
	Agent        map[string]any
	TraceID      string
	StartedAt    string
	Events       []map[string]any
	ToolManifest []map[string]any
	Redact       func(any) any
}

func NewHarness(project string, agent map[string]any, traceID string, startedAt string) *Harness {
	if traceID == "" {
		traceID = "tr_go_conformance"
	}
	if startedAt == "" {
		startedAt = now()
	}
	return &Harness{
		Project:      project,
		Agent:        agent,
		TraceID:      traceID,
		StartedAt:    startedAt,
		Events:       []map[string]any{},
		ToolManifest: []map[string]any{},
		Redact:       DefaultRedactor,
	}
}

func (h *Harness) RecordInput(input any) {
	h.Events = append(h.Events, stripNil(map[string]any{
		"id":    eventID(len(h.Events)),
		"type":  "user_input",
		"ts":    now(),
		"input": h.Redact(input),
	}))
}

func (h *Harness) RecordToolCall(tool string, args map[string]any, response any, approval map[string]any, preState any, postState any, sideEffects []map[string]any, durationMs int) {
	safeArgs := h.Redact(args)
	safeResponse := h.Redact(response)
	h.Events = append(h.Events, stripNil(map[string]any{
		"id":           eventID(len(h.Events)),
		"type":         "tool_call",
		"ts":           now(),
		"tool":         tool,
		"args":         safeArgs,
		"argHash":      SHA256(safeArgs),
		"response":     safeResponse,
		"responseHash": SHA256(safeResponse),
		"approval":     redactOptional(h.Redact, approval),
		"preState":     redactOptional(h.Redact, preState),
		"postState":    redactOptional(h.Redact, postState),
		"sideEffects":  h.Redact(sideEffects),
		"durationMs":   durationMs,
	}))
}

func (h *Harness) RecordFinalOutput(output any) {
	h.Events = append(h.Events, stripNil(map[string]any{
		"id":     eventID(len(h.Events)),
		"type":   "final_output",
		"ts":     now(),
		"output": h.Redact(output),
	}))
}

func (h *Harness) Finalize(expectedOutcome map[string]any, metadata map[string]any) map[string]any {
	return stripNil(map[string]any{
		"schemaVersion":    "agentreplay.trace.v1",
		"traceId":          h.TraceID,
		"project":          h.Project,
		"agent":            h.Agent,
		"startedAt":        h.StartedAt,
		"endedAt":          now(),
		"replayMismatches": []any{},
		"toolManifest":     h.ToolManifest,
		"events":           h.Events,
		"expectedOutcome":  expectedOutcome,
		"metadata":         metadata,
	})
}

func (h *Harness) Save(path string, expectedOutcome map[string]any, metadata map[string]any) error {
	trace := h.Finalize(expectedOutcome, metadata)
	bytes, err := json.MarshalIndent(trace, "", "  ")
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		return err
	}
	return os.WriteFile(path, append(bytes, '\n'), 0644)
}

func DefaultRedactor(value any) any {
	switch typed := value.(type) {
	case string:
		if defaultEmailPattern.MatchString(typed) {
			return "[REDACTED]"
		}
		return typed
	case []map[string]any:
		out := make([]map[string]any, len(typed))
		for i, item := range typed {
			out[i] = DefaultRedactor(item).(map[string]any)
		}
		return out
	case []any:
		out := make([]any, len(typed))
		for i, item := range typed {
			out[i] = DefaultRedactor(item)
		}
		return out
	case map[string]any:
		out := map[string]any{}
		for key, item := range typed {
			if defaultKeyPattern.MatchString(key) {
				out[key] = "[REDACTED]"
			} else {
				out[key] = DefaultRedactor(item)
			}
		}
		return out
	default:
		return value
	}
}

func SHA256(value any) string {
	sum := sha256.Sum256([]byte(StableStringify(value)))
	return hex.EncodeToString(sum[:])
}

func StableStringify(value any) string {
	bytes, _ := json.Marshal(sortJSON(value))
	return string(bytes)
}

func sortJSON(value any) any {
	switch typed := value.(type) {
	case map[string]any:
		keys := make([]string, 0, len(typed))
		for key := range typed {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		out := map[string]any{}
		for _, key := range keys {
			out[key] = sortJSON(typed[key])
		}
		return out
	case []map[string]any:
		out := make([]any, len(typed))
		for i, item := range typed {
			out[i] = sortJSON(item)
		}
		return out
	case []any:
		out := make([]any, len(typed))
		for i, item := range typed {
			out[i] = sortJSON(item)
		}
		return out
	default:
		return value
	}
}

func stripNil(value map[string]any) map[string]any {
	out := map[string]any{}
	for key, item := range value {
		if item != nil {
			out[key] = item
		}
	}
	return out
}

func redactOptional(redact func(any) any, value any) any {
	if value == nil {
		return nil
	}
	return redact(value)
}

func now() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05.000Z")
}

func eventID(index int) string {
	return "evt_" + leftPad(index+1, 4)
}

func leftPad(value int, width int) string {
	text := "0000000000" + itoa(value)
	return text[len(text)-width:]
}

func itoa(value int) string {
	if value == 0 {
		return "0"
	}
	digits := []byte{}
	for value > 0 {
		digits = append([]byte{byte('0' + value%10)}, digits...)
		value /= 10
	}
	return string(digits)
}
